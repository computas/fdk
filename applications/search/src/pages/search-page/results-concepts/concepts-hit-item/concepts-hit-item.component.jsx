import React from 'react';
import PropTypes from 'prop-types';
import * as _ from 'lodash';

import localization from '../../../../lib/localization';
import { getTranslateText } from '../../../../lib/translateText';
import { PublisherLabel } from '../../../../components/publisher-label/publisher-label.component';
import './concepts-hit-item.scss';

const renderPublisher = source => {
  const { creator } = source;

  if (!creator) {
    return null;
  }
  return <PublisherLabel publisherItem={creator} />;
};

const renderThemes = source => {
  const { inScheme } = source;
  const children = items =>
    items.map((item, index) => {
      const subItem = item.substring(item.lastIndexOf('/') + 1);
      return (
        <span
          key={`dataset-description-inScheme-${index}`}
          className="fdk-label"
        >
          <span className="uu-invisible" aria-hidden="false">
            Tema.
          </span>
          {subItem}
        </span>
      );
    });
  if (inScheme) {
    return <div className="mt-3">{children(inScheme)}</div>;
  }
  return null;
};

const renderLaw = _source => {
  const { source } = _source;
  if (source) {
    return (
      <div>
        <span className="fa-stack fdk-fa-left fdk-fa-circle">
          <i className="fa fa-file-text fa-stack-1x fdk-color0" />
        </span>
        <a title="Link til lovhjemmel for begrep" href={source}>
          <span className="uu-invisible" aria-hidden="false">
            Link til lovhjemmel for begrep.
          </span>
          {source}
          <i className="fa fa-external-link fdk-fa-right" />
        </a>
      </div>
    );
  }
  return null;
};

const renderNote = source => {
  const { note } = source;
  if (note) {
    return <p className="fdk-p-search-hit">{getTranslateText(note)}</p>;
  }
  return null;
};

const renderAltLabel = source => {
  const { altLabel } = source;
  const children = items =>
    items.map((item, index) => {
      if (index > 0) {
        return (
          <span key={`concepts-altlabel-${index}`}>
            {`, ${getTranslateText(item)}`}
          </span>
        );
      }
      return (
        <span key={`concepts-altlabel-${index}`}>
          {`${getTranslateText(item)}`}
        </span>
      );
    });
  if (altLabel) {
    return (
      <p>
        <span className="uu-invisible" aria-hidden="false">
          Begrep er
        </span>
        <strong>{localization.terms.altLabel} </strong>
        {children(altLabel)}
      </p>
    );
  }
  return null;
};

const renderDocCount = result => {
  const { _source } = result;

  const subjectCountItem = _source.datasets ? _source.datasets.length : 0;
  if (subjectCountItem > 0 && _source.prefLabel) {
    return (
      <p>
        <a
          className="fdk-hit-dataset-count"
          title="Link til datasett med begrep"
          href={`/?subject=${getTranslateText(_source.prefLabel)}`}
        >
          {localization.terms.docCount} {subjectCountItem}{' '}
          {localization.terms.docCountPart2}
        </a>
      </p>
    );
  }
  return null;
};

export const ConceptsHitItem = props => {
  const { onAddTerm } = props;
  const { _source } = props.result;
  const { prefLabel, definition, uri } = _source;

  let termTitle;
  let termDescription;

  if (prefLabel) {
    termTitle = getTranslateText(prefLabel);
    termTitle =
      termTitle.charAt(0).toUpperCase() + termTitle.substring(1).toLowerCase();
  }
  if (definition) {
    termDescription = getTranslateText(definition);
  }

  let toBeCompared = false;
  if (props.terms) {
    toBeCompared = _.some(props.terms, term => term.uri === uri);
  }

  return (
    <article
      className="fdk-a-search-hit"
      title={`Begrep: ${termTitle}`}
      tabIndex="0" // eslint-disable-line jsx-a11y/no-noninteractive-tabindex
    >
      <span className="uu-invisible" aria-hidden="false">
        SÃ¸ketreff begrep.
      </span>
      <div
        className={`fdk-container-search-hit ${
          toBeCompared ? 'toBeCompared' : ''
        }`}
      >
        {!toBeCompared && (
          <button
            className="btn btn-primary fdk-button float-right mt-3 d-none d-lg-inline"
            onClick={() => {
              onAddTerm(_source);
            }}
            type="button"
          >
            <span aria-hidden="true">+</span>
            {localization.compare.addCompare}
          </button>
        )}

        {!toBeCompared && (
          <button
            className="fdk-button fdk-button-default fdk-btn-compare d-block d-lg-none"
            onClick={() => {
              onAddTerm(_source);
            }}
            type="button"
          >
            <span aria-hidden="true">+</span>
            {localization.compare.addCompare}
          </button>
        )}

        <span className="uu-invisible" aria-hidden="false">
          Tittel.
        </span>
        <h2 className="inline-block mr-2">{termTitle}</h2>

        {renderPublisher(_source)}

        {renderThemes(_source)}

        <p className="fdk-p-search-hit">
          <span className="uu-invisible" aria-hidden="false">
            Beskrivelse av begrep.
          </span>
          {termDescription}
        </p>

        {renderLaw(_source)}

        <hr />

        {renderNote(_source)}

        {renderAltLabel(_source)}

        {renderDocCount(props.result)}
      </div>
    </article>
  );
};

ConceptsHitItem.defaultProps = {
  result: null,
  terms: null
};

ConceptsHitItem.propTypes = {
  result: PropTypes.shape({}),
  terms: PropTypes.array,
  onAddTerm: PropTypes.func.isRequired
};
