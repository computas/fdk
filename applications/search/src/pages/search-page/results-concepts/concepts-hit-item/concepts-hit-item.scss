@import '../../../../../node_modules/designsystemet/fdk-designsystem-bootstrap4/scss/colors.scss';

.fdk-a-search-hit {
  text-decoration: none;
}

.fdk-fa-circle {
  background-color: $color1;
  -webkit-border-radius: 50%;
  -moz-border-radius: 50%;
  border-radius: 50%;
}

.toBeCompared {
  border: 2px solid $color1 !important;
}

.fdk-btn-compare {
  margin: 0 0 1.5em 0 !important;
}