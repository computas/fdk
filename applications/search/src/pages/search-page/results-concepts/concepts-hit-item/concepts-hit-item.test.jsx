import React from 'react';
import { shallow } from 'enzyme';
import { ConceptsHitItem } from './concepts-hit-item.component';
import concepts from '../../../../../test/fixtures/concepts';

let onAddTerm;
let defaultProps;
let wrapper;

beforeEach(() => {
  onAddTerm = jest.fn();
  defaultProps = {
    result: concepts.hits.hits[0],
    terms: null,
    onAddTerm
  };
  wrapper = shallow(<ConceptsHitItem {...defaultProps} />);
});

test('should render ConceptsHitItem correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onAddTerm', () => {
  const shallowWrapper = shallow(<ConceptsHitItem {...defaultProps} />);
  shallowWrapper.find('button.d-lg-inline').prop('onClick')(
    concepts.hits.hits[0]._source
  );
  expect(onAddTerm).toHaveBeenLastCalledWith(concepts.hits.hits[0]._source);
  shallowWrapper.find('button.d-lg-inline').prop('onClick')(
    concepts.hits.hits[1]._source
  );
  expect(onAddTerm).toHaveBeenLastCalledWith(concepts.hits.hits[0]._source);
});

test('should render ConceptsHitItem correctly with compare button not showing', () => {
  const props = {
    result: concepts.hits.hits[0],
    terms: [concepts.hits.hits[0]._source],
    onAddTerm
  };
  const shallowWrapper = shallow(<ConceptsHitItem {...props} />);
  expect(shallowWrapper).toMatchSnapshot();
});

test('should render ConceptsHitItem correctly with no alt label, no source and no note', () => {
  const props = {
    result: concepts.hits.hits[1],
    terms: null,
    onAddTerm
  };
  const shallowWrapper = shallow(<ConceptsHitItem {...props} />);
  expect(shallowWrapper).toMatchSnapshot();
});
