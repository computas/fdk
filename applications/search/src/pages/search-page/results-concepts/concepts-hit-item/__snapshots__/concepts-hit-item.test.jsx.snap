// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ConceptsHitItem correctly 1`] = `
<article
  className="fdk-a-search-hit"
  tabIndex="0"
  title="Begrep: Enhet"
>
  <span
    aria-hidden="false"
    className="uu-invisible"
  >
    Søketreff begrep.
  </span>
  <div
    className="fdk-container-search-hit "
  >
    <button
      className="btn btn-primary fdk-button float-right mt-3 d-none d-lg-inline"
      onClick={[Function]}
      type="button"
    >
      <span
        aria-hidden="true"
      >
        +
      </span>
      Add to compare
    </button>
    <button
      className="fdk-button fdk-button-default fdk-btn-compare d-block d-lg-none"
      onClick={[Function]}
      type="button"
    >
      <span
        aria-hidden="true"
      >
        +
      </span>
      Add to compare
    </button>
    <span
      aria-hidden="false"
      className="uu-invisible"
    >
      Tittel.
    </span>
    <h2
      className="inline-block mr-2"
    >
      Enhet
    </h2>
    <PublisherLabel
      label={null}
      publisherItem={
        Object {
          "id": "974760673",
          "naeringskode": Object {
            "code": "84.110",
            "prefLabel": Object {
              "no": "Generell offentlig administrasjon",
            },
            "uri": "http://www.ssb.no/nace/sn2007/84.110",
          },
          "name": "Brønnøysundregistrene",
          "orgPath": "/STAT/912660680/974760673",
          "organisasjonsform": "ORGL",
          "overordnetEnhet": "912660680",
          "sektorkode": Object {
            "code": "6100",
            "prefLabel": Object {
              "no": "Statsforvaltningen",
            },
            "uri": "http://www.brreg.no/sektorkode/6100",
          },
          "type": "no.dcat.datastore.domain.dcat.Publisher",
          "uri": "http://data.brreg.no/enhetsregisteret/enhet/974760673",
          "valid": true,
        }
      }
    />
    <div
      className="mt-3"
    >
      <span
        className="fdk-label"
        key="dataset-description-inScheme-0"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Tema.
        </span>
        Befolkning
      </span>
    </div>
    <p
      className="fdk-p-search-hit"
    >
      <span
        aria-hidden="false"
        className="uu-invisible"
      >
        Beskrivelse av begrep.
      </span>
      alt som er registrert med et organisasjonsnummer 
    </p>
    <div>
      <span
        className="fa-stack fdk-fa-left fdk-fa-circle"
      >
        <i
          className="fa fa-file-text fa-stack-1x fdk-color0"
        />
      </span>
      <a
        href="https://jira.brreg.no/browse/BEGREP-208"
        title="Link til lovhjemmel for begrep"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Link til lovhjemmel for begrep.
        </span>
        https://jira.brreg.no/browse/BEGREP-208
        <i
          className="fa fa-external-link fdk-fa-right"
        />
      </a>
    </div>
    <hr />
    <p
      className="fdk-p-search-hit"
    >
      Alle hovedenheter, underenheter og organisasjonsledd som er identifisert med et organisasjonsnummer.
    </p>
    <p>
      <a
        className="fdk-hit-dataset-count"
        href="/?subject=enhet"
        title="Link til datasett med begrep"
      >
        Used in
         
        1
         
        dataset
      </a>
    </p>
  </div>
</article>
`;

exports[`should render ConceptsHitItem correctly with compare button not showing 1`] = `
<article
  className="fdk-a-search-hit"
  tabIndex="0"
  title="Begrep: Enhet"
>
  <span
    aria-hidden="false"
    className="uu-invisible"
  >
    Søketreff begrep.
  </span>
  <div
    className="fdk-container-search-hit toBeCompared"
  >
    <span
      aria-hidden="false"
      className="uu-invisible"
    >
      Tittel.
    </span>
    <h2
      className="inline-block mr-2"
    >
      Enhet
    </h2>
    <PublisherLabel
      label={null}
      publisherItem={
        Object {
          "id": "974760673",
          "naeringskode": Object {
            "code": "84.110",
            "prefLabel": Object {
              "no": "Generell offentlig administrasjon",
            },
            "uri": "http://www.ssb.no/nace/sn2007/84.110",
          },
          "name": "Brønnøysundregistrene",
          "orgPath": "/STAT/912660680/974760673",
          "organisasjonsform": "ORGL",
          "overordnetEnhet": "912660680",
          "sektorkode": Object {
            "code": "6100",
            "prefLabel": Object {
              "no": "Statsforvaltningen",
            },
            "uri": "http://www.brreg.no/sektorkode/6100",
          },
          "type": "no.dcat.datastore.domain.dcat.Publisher",
          "uri": "http://data.brreg.no/enhetsregisteret/enhet/974760673",
          "valid": true,
        }
      }
    />
    <div
      className="mt-3"
    >
      <span
        className="fdk-label"
        key="dataset-description-inScheme-0"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Tema.
        </span>
        Befolkning
      </span>
    </div>
    <p
      className="fdk-p-search-hit"
    >
      <span
        aria-hidden="false"
        className="uu-invisible"
      >
        Beskrivelse av begrep.
      </span>
      alt som er registrert med et organisasjonsnummer 
    </p>
    <div>
      <span
        className="fa-stack fdk-fa-left fdk-fa-circle"
      >
        <i
          className="fa fa-file-text fa-stack-1x fdk-color0"
        />
      </span>
      <a
        href="https://jira.brreg.no/browse/BEGREP-208"
        title="Link til lovhjemmel for begrep"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Link til lovhjemmel for begrep.
        </span>
        https://jira.brreg.no/browse/BEGREP-208
        <i
          className="fa fa-external-link fdk-fa-right"
        />
      </a>
    </div>
    <hr />
    <p
      className="fdk-p-search-hit"
    >
      Alle hovedenheter, underenheter og organisasjonsledd som er identifisert med et organisasjonsnummer.
    </p>
    <p>
      <a
        className="fdk-hit-dataset-count"
        href="/?subject=enhet"
        title="Link til datasett med begrep"
      >
        Used in
         
        1
         
        dataset
      </a>
    </p>
  </div>
</article>
`;

exports[`should render ConceptsHitItem correctly with no alt label, no source and no note 1`] = `
<article
  className="fdk-a-search-hit"
  tabIndex="0"
  title="Begrep: Organisasjonsnummer"
>
  <span
    aria-hidden="false"
    className="uu-invisible"
  >
    Søketreff begrep.
  </span>
  <div
    className="fdk-container-search-hit "
  >
    <button
      className="btn btn-primary fdk-button float-right mt-3 d-none d-lg-inline"
      onClick={[Function]}
      type="button"
    >
      <span
        aria-hidden="true"
      >
        +
      </span>
      Add to compare
    </button>
    <button
      className="fdk-button fdk-button-default fdk-btn-compare d-block d-lg-none"
      onClick={[Function]}
      type="button"
    >
      <span
        aria-hidden="true"
      >
        +
      </span>
      Add to compare
    </button>
    <span
      aria-hidden="false"
      className="uu-invisible"
    >
      Tittel.
    </span>
    <h2
      className="inline-block mr-2"
    >
      Organisasjonsnummer
    </h2>
    <p
      className="fdk-p-search-hit"
    >
      <span
        aria-hidden="false"
        className="uu-invisible"
      >
        Beskrivelse av begrep.
      </span>
      Et nisifret nummer som tildeles av Enhetsregisteret for en organisasjon som skal operere som en offentlig aktør
    </p>
    <hr />
    <p>
      <span
        aria-hidden="false"
        className="uu-invisible"
      >
        Begrep er
      </span>
      <strong>
        Også kjent som:
         
      </strong>
      <span
        key="concepts-altlabel-0"
      >
        Orgid
      </span>
      <span
        key="concepts-altlabel-1"
      >
        , Orgnr
      </span>
    </p>
    <p>
      <a
        className="fdk-hit-dataset-count"
        href="/?subject=organisasjonsnummer"
        title="Link til datasett med begrep"
      >
        Used in
         
        8
         
        dataset
      </a>
    </p>
  </div>
</article>
`;
